mode_calculations.py

```diff
@@ -600,3 +601,366 @@ def inner_product(t, abar, b, axis=None, apply_conjugate=False):
+        which range over ell, ellp values up to and including ell_max,
+        for only the non-vanishing matrix elements of M.
+
+    b : WaveformModes object
+        The waveform |b>. This function is linear in `b`.
+
+    allow_LM_differ : bool, optional [default: False]
+        If True and if the set of (ell,m) modes between a and b
+        differ, then the inner product will be computed using the
+        intersection of the set of modes.
+
+    allow_times_differ: bool, optional [default: False]
+        If True and if the set of times between a and b differ,
+        then both WaveformModes will be interpolated to the
+        intersection of the set of times.
+
+    Returns
+    -------
+    times, expect_val : ndarray, complex ndarray
+        Resulting tuple has length two.  The first element is the set
+        of times u for the timeseries.  The second element is the
+        timeseries of <a|M|b>(u).
+
+    """
+
+    from .extrapolation import intersection
+    from spherical_functions import LM_index
+
+    if (a.spin_weight != b.spin_weight):
+        raise ValueError("Spin weights must match in matrix_expect_val")
+
+    LM_clip = slice(a.ell_min, a.ell_max + 1)
+    if ((a.ell_min != b.ell_min) or (a.ell_max != b.ell_max)):
+        if (allow_LM_differ):
+            LM_clip = slice( max(a.ell_min, b.ell_min),
+                             min(a.ell_max, b.ell_max) + 1 )
+            if (LM_clip.start >= LM_clip.stop):
+                raise ValueError("Intersection of (ell,m) modes is "
+                                 "empty.  Assuming this is not desired.")
+        else:
+            raise ValueError("ell_min and ell_max must match in matrix_expect_val "
+                             "(use allow_LM_differ=True to override)")
+
+    t_clip = None
+    if not np.array_equal(a.t, b.t):
+        if (allow_times_differ):
+            t_clip = intersection(a.t, b.t)
+        else:
+            raise ValueError("Time samples must match in matrix_expect_val "
+                             "(use allow_times_differ=True to override)")
+
+    ##########
+
+    times = a.t
+    A = a
+    B = b
+
+    if (LM_clip is not None):
+        A = A[:,LM_clip]
+        B = B[:,LM_clip]
+
+    if (t_clip is not None):
+        times = t_clip
+        A = A.interpolate(t_clip)
+        B = B.interpolate(t_clip)
+
+    ##########
+    ## Time for actual math!
+
+    Abar_data  = np.conj(A.data)
+    B_data     = B.data
+
+    expect_val = np.zeros(times.size, dtype=np.complex)
+
+    ell_min = LM_clip.start
+    ell_max = LM_clip.stop - 1
+
+    for ellp, mp, ell, m, M_el in M(ell_min, ell_max):
+        # QUESTION: Should we test if M_el is very close to 0, and
+        # continue to the next element if it's close enough?
+
+        i_a = LM_index(ellp, mp, A.ell_min)
+        i_b = LM_index(ell , m , B.ell_min)
+
+        # QUESTION: Do any python implementations short-circuit the
+        # multiplication if M_el is 0?
```

Comment:
Alright, then it would be incumbent upon anybody who creates one of these sparse matrices to do their due diligence and only emit nonzero matrix elements.